### DISCUSSION TITLE
SD card Configuration

### DISCUSSION THREAD

**Discussion by Sudupatil** [Bodmer/TFT_eSPI#2717] [Q&A - General]
Hi,
 I am using ILI9488  3.5 Tft Display. I would like to use SD card with it. but i don't know how. could you please help me with it. what are the configuration setting i have to follow.
LED	3v+
SCK	18
SDI(MOSI)	23
DC	4
RESET	2
CS	15
GND	3V -
VCC	3V+

SD_CS	5
SD_MOSI	18
SD_MISO	19
[Setup20_ILI9488.zip](https://github.com/Bodmer/TFT_eSPI/files/11817077/Setup20_ILI9488.zip)

SD_SCK	23

From One of discussion "https://forum.arduino.cc/t/sd-card-with-tft-espi/1023970/9" it's been conclude that SD card Module can't use.

setup configuration files are attached.

Thanks and regards
Sudarshan

**Comment by TangerineDreamer** [comment]
Hi, sudupatil,

i've seen this question at a lot of place in the web,i was like you,and i could collect pieces in the forum to build my settings.
I use the "mandatory" pinout to have the maximum frequency transmissions using  the TFT and Sd card.
The setup:

```
//TFT & SD CARD & ESP32 (2 SPI MODULES)

//#define ST7796_DRIVER
//#define ILI9341_DRIVER
#define ILI9488_DRIVER

//TFT Pins (HSPI)
#define TFT_MISO 12
#define TFT_MOSI 13
#define TFT_SCLK 14
#define TFT_CS   15  // Chip select control pin//choose what you want
#define TFT_DC    2  // Data Command control pin//choose what you want
#define TFT_RST   4  // Reset pin (or could connect to RST pin if you want)
//#define TFT_RST  -1  // Set TFT_RST to -1 ( if display RESET is connected to ESP32 board RST)



//Touch Pins (HSPI)
#define TOUCH_DO 12 
#define TOUCH_DIN 13 
#define TOUCH_CLK 14
#define TOUCH_CS 33//27 // Chip select pin (T_CS) of touch screen//choose what you want


//SD Card Pins (VSPI)
#define SD_MISO 19 //SD PINS
#define SD_MOSI 23
#define SD_SCLK 18
#define SD_CS   5   //i prefered to use this pin             




#define LOAD_GLCD   // Font 1. Original Adafruit 8 pixel font needs ~1820 bytes in FLASH
#define LOAD_FONT2  // Font 2. Small 16 pixel high font, needs ~3534 bytes in FLASH, 96 characters
#define LOAD_FONT4  // Font 4. Medium 26 pixel high font, needs ~5848 bytes in FLASH, 96 characters
#define LOAD_FONT6  // Font 6. Large 48 pixel font, needs ~2666 bytes in FLASH, only characters 1234567890:-.apm
#define LOAD_FONT7  // Font 7. 7 segment 48 pixel font, needs ~2438 bytes in FLASH, only characters 1234567890:-.
#define LOAD_FONT8  // Font 8. Large 75 pixel font needs ~3256 bytes in FLASH, only characters 1234567890:-.
//#define LOAD_FONT8N // Font 8. Alternative to Font 8 above, slightly narrower, so 3 digits fit a 160 pixel TFT
#define LOAD_GFXFF  // FreeFonts. Include access to the 48 Adafruit_GFX free fonts FF1 to FF48 and custom fonts


#define SMOOTH_FONT



#define SPI_FREQUENCY  20000000

#define SPI_READ_FREQUENCY  20000000


#define SPI_TOUCH_FREQUENCY  2500000

#define USE_HSPI_PORT
```

The sketch BASE after the predefinitions:

```
//#include blabla...
SPIClass spiSD(VSPI);
SPIClass spiTFT(HSPI);

const int SD_CS_PIN = 5;//for the sd or sdfat stuff
```
The setup:

```
Serial.begin(115200);//if you want monitor debug
spiSD.begin(18,19,23,5);
spiTFT.begin(14,12,13,15);
//...and so on
```

**Reply by Sparky54** [reply]
Hi, I think I have a very similar issue. I am looking to use a TFT display with a sd card module also, but when I initialize the TFT display, the serial monitor stops receiving data from the sd card module, which I think tells me that there is a conflict between the two SPI setups. I've tried troubleshooting the issue by adjusting the code and circuit configuration, but I haven't been successful in solving the problem without commenting out the TFT initialization.

These are the pin assignments for each module:

TFT Display:

MOSI: Pin 13
SCLK: Pin 14
CS: Pin 15
DC: Pin 22
RST: Pin 21

SD Card Module:
CS: Pin 22
MISO: Pin 19
MOSI: Pin 23
CLK: Pin 18



I suspect that the conflict arises from conflicting SPI bus settings. I setup the custom SPI pins for the sd card module in my program by doing the following: 

#include <SPI.h>
#define PIN_SD_CARD_CS 22
#define PIN_SD_CARD_MISO 19
#define PIN_SD_CARD_MOSI 23
#define PIN_SD_CARD_CLK 18

void setup {
  SPI.begin(PIN_SD_CARD_CLK, PIN_SD_CARD_MISO, PIN_SD_CARD_MOSI, PIN_SD_CARD_CS);
}

...and the TFT pins were of course set up in the user_setup file. 

I'm wondering if this solution could be applied to my code and was unclear how it worked. You setup the sd card pins in the user_setup file instead??

**Reply by ctc-nick** [reply]
This helps alot and works also for ESP32_2432S028_R. Thank you! :relaxed:

**Comment by TangerineDreamer** [comment]
Hi, Sparky54,
I suppose it was a question for me concerning my intervention :)
i believe you can setup the sdcard outside the user setup,and directly in the sketch,as the usersetup only concern the tft display.
You were right about the SPI.begin(PIN_SD_CARD_CLK, PIN_SD_CARD_MISO, PIN_SD_CARD_MOSI, PIN_SD_CARD_CS);
i forgot to precise what it meant.
I post a sd card & tft test to verify quickly if that works for you:

```
#include "FS.h"
#include <SPI.h>
#include "SdFat.h"
#include "FreeStack.h"

SPIClass spiSD(VSPI);
SPIClass spiTFT(HSPI);
// **************** SD CARD
const int SD_CS_PIN = 5;
#define SPI_SPEED SD_SCK_MHZ(10)
// SD_FAT_TYPE = 0 for SdFat/File as defined in SdFatConfig.h,
// 1 for FAT16/FAT32, 2 for exFAT, 3 for FAT16/FAT32 and exFAT.
#define SD_FAT_TYPE 0

#if SD_FAT_TYPE == 0
SdFat sd;
File dir;
File file;
#elif SD_FAT_TYPE == 1
SdFat32 sd;
File32 dir;
File32 file;
#elif SD_FAT_TYPE == 2
SdExFat sd;
ExFile dir;
ExFile file;
#elif SD_FAT_TYPE == 3
SdFs sd;
FsFile dir;
FsFile file;
#else  // SD_FAT_TYPE
#error invalid SD_FAT_TYPE
#endif  // SD_FAT_TYPE

void setup() {

  spiSD.begin(18,19,23,5);
spiTFT.begin(14,12,13,15);

  Serial.begin(115200);
  
  while (!Serial);
  Serial.println("SD Card directory sorting...");
  if (!sd.begin(SD_CS_PIN, SPI_SPEED)) {
    Serial.println("SD card error!");
    while (true);
  }
  sd.ls("/", LS_R);
}
void loop() {
  
}
```
Freestack.h is in the Sdfat library

**Comment by TangerineDreamer** [comment]
Sorry but i forgot to post an example with tft espi displaying results:

```

#include "FS.h"
#include "SD.h"
#include "SPI.h"

SPIClass spiSD(VSPI);
SPIClass spiTFT(HSPI);

#include <TFT_eSPI.h>      // Hardware-specific library

TFT_eSPI tft = TFT_eSPI(); // Invoke custom library

File root;

int fileCountOnSD = 0; // for counting files



void setup(){
    Serial.begin(115200);

spiSD.begin(18,19,23,5);
spiTFT.begin(14,12,13,15);

// Initialise the TFT screen
  tft.init();

  // Set the rotation before we calibrate
  tft.setRotation(0);
 
  // Clear the screen
  tft.fillScreen(TFT_BLACK);

  // Draw keypad background
  tft.fillRect(0, 0, 240, 320, TFT_DARKGREY);//adapt it to your screen 

  tft.setTextDatum(TL_DATUM);        // Use top left corner as text coord datum
      tft.setFreeFont(&FreeSans9pt7b);  // Choose a nicefont that fits box
      tft.setTextColor(TFT_WHITE);     // Set the font colour

 while (!Serial) {
    ; // wait for serial port to connect. Needed for native USB port only
  }

  Serial.print("Initializing SD card...");

  if (!SD.begin()) {
    Serial.println("initialization failed!");
    while (1);
  }
  Serial.println("initialization done.");

  root = SD.open("/");

  printDirectory(root, 0);

  // Now print the total files count
  Serial.println(F("fileCountOnSD: "));
  Serial.println(fileCountOnSD);

  tft.drawNumber(fileCountOnSD, 10, 30);
  

  Serial.println("done!");

}

void loop(){

}

void printDirectory(File dir, int numTabs) {

   tft.setTextDatum(TL_DATUM);        // Use top left corner as text coord datum
      tft.setFreeFont(&FreeSans9pt7b);  // Choose a nicefont that fits box
      tft.setTextColor(TFT_WHITE);     // Set the font colour
      
  while (true) {

    File entry =  dir.openNextFile();
    if (! entry) {
      // no more files
      break;
    }
    for (uint8_t i = 0; i < numTabs; i++) {
      Serial.print('\t');
    }
    Serial.print(entry.name());
     tft.drawString(entry.name(), 10, 10);
    

    // for each file count it
    fileCountOnSD++;

    if (entry.isDirectory()) {
      Serial.println("/");
      printDirectory(entry, numTabs + 1);
    } else {
      // files have sizes, directories do not
      Serial.print("\t\t");
      Serial.println(entry.size(), DEC);
    }
    entry.close();
  }
}

```

**Comment by rradomir** [comment]
In my case, screen with built-in sd-card reader works with this:


```
tft.init();

digitalWrite(SD_CS,HIGH);
digitalWrite(TFT_CS,HIGH);
delay(10);
SD.begin(SD_CS, tft.getSPIinstance());
```
Of course all SPI pins are the same, only SD_CS is additional connection besides screen pins.

**Reply by Sudupatil** [reply]
![PCB_NEW_PCB15_8_2023-06-16](https://github.com/Bodmer/TFT_eSPI/assets/99311112/2b40971b-1224-4132-ba89-0391158652fd)
![IMG_20230529_081050](https://github.com/Bodmer/TFT_eSPI/assets/99311112/636ba0fd-324f-477c-9f5a-e5cb66947aed)
Hello Bodmer and all,
[Connection.xlsx](https://github.com/Bodmer/TFT_eSPI/files/12060610/Connection.xlsx)

Please go through attached excel file where I have mentioned my current ESP32 PIN configuration.
ESP 32 has some common pins, I stated what’s working and what is intended.

Will my configuration work?  

If possible, then please advise on whether the MAX31865 configuration will work or not.

I don't have my PCB and setup with me. So i can't test. I have to get it from Customer (if required).

**Reply by vealeal** [reply]
This worked for me too

**Comment by Sudupatil** [comment]
Hello Every one,

i have done the suggested changes, SD card is configured and working, but TFT is not working accordingly. 

  tft.fillScreen(MainColor);
  tft.setTextSize(1);
  tft.setTextDatum(TC_DATUM);
  tft.drawString("ALERT", 240, 80, 6);
![IMG_20230803_122738](https://github.com/Bodmer/TFT_eSPI/assets/99311112/6d5fbd58-bf18-40bb-9d0e-b75745c9030d)

  tft.drawString("initialization done", 240 , 140, 6); //padding + 20

the above code should Display Alert in middle and below line should display initialization done.  but it's showing something like as attached. can anyone tell me why?

**Reply by shetyespratham** [reply]
Can you please post a complete working example of SDCARD along with TFT display?  I tried whatever mentioned in this post, nothing worked for me.  I am okay with sharing the same SPI lines with TFT display or having a different set of pins for SDCARD SPI.  Other things are working pretty fine.  Struggling only with getting SD card working.

**Comment by TangerineDreamer** [comment]
I'm sorry for that,
did you fill the setup file (not the one in the sketch) with the right settings?
did you insert a sd card in the player?
do you see something in the monitor to debug?
Sometimes the sd card handling is exclusive and all is blocked if you don't have a sd card in,or maybe your
sd card is not well formatted,and have you tried with a 16Gb (not 32) one?

**Reply by shetyespratham** [reply]
No Problem.

I tested all connections and found that somehow the MISO pin of SD card socket is not connected to MISO pin of SD header.  I soldered it and then finally everything is working.  No need of VSPI and all.  Everything i.e. TFT, Touch and SD Card all working on shared SPI pins.  (only TFT MISO I have kept unconnected).  I am working on ESP32-S3. So found that it is working with any custom pins without any problem.

Thank you so much.
